### PR TITLE
Add support for meson build system

### PR DESCRIPTION
**Pull request by martin-belanger**
Hi @keithbusch 

As we discussed earlier this week, this is the pull request that adds support for the meson build system. As you will see, I did not modify any of the existing Makefiles. This simply adds meson in parallel to the existing Makefiles. 

I added text to the README file to explain how to use meson to build libnvme. I also gave a lot of examples showing how one can change the build configuration from the command line w/o having to modify any files.

Please review at your convenience. If you have any questions, please don't hesitate to contact me. I can also give you a live demo over Zoom if you'd like.

Regards,

Martin Belanger
Technical Staff / Engineering Technologist
Dell Inc.

### CODE REVIEW

**Comment by birkelund** [comment]
Hi @martin-belanger 

For the CCAN configurator - I know your pain.

I have another project where I use CCAN stuff, and the way I did this in meson is to use a custom target in `ccan/meson.build`:

```meson
subdir('tools/configurator')

ccan_config_h = custom_target('ccan_config_h',
  output: 'config.h',
  command: [ccan_configurator],
  capture: true,
  build_by_default: true,
)

ccan_sources = [
  'ccan/err/err.c',
  'ccan/list/list.c',
  'ccan/pr_log/pr_log.c',
  'ccan/str/str.c',
  'ccan/time/time.c',
]

ccan_lib = static_library('ccan',
  [ccan_config_h, ccan_sources],
  include_directories : ccan_inc,
)
```

In `ccan/tools/configurator/meson.build`:

```meson
ccan_configurator = executable('configurator',
  files('configurator.c'),
  c_args: ['-Wno-missing-field-initializers'],
)
```

In the top-level `meson.build` I have

```meson
ccan_inc = include_directories('ccan')
subdir('ccan')
```

Then, in `src/meson.build` I just reference this

```meson
my_lib = library('my_lib', [ccan_config_h, my_sources],
  include_directories: [my_lib_inc, ccan_inc],
  dependencies: ...,
  link_with: [ccan_lib],
)
```

**Comment by birkelund** [comment]
Oh sorry, I missed that that is exactly how you do it - I only saw your attempt at replacing it... Nevermind the noise then :)

**Comment by martin-belanger** [comment]
No problem @birkelund . As you saw, I tried to resolve the CCAN with an all-meson approach, but I gave up and simply used a custom target. I may come back to it later to see if I can eliminate the custom target completely. The thing is that we are not using all that the CCAN code is testing for. So it could be simplified.

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+project(
+    'libnvme',
+    'c', 'cpp',
+    version: '0.1',
+    license: 'LGPLv2+',
+    default_options: [
+        'buildtype=release',
+        'prefix=/usr',
+    ]
+)
```
Comment (verschwindibussen) [inline meson.build:70]: Do we need a minimum version for meson here?

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+configure_file(
+    output: 'config-host.h',
+    configuration: conf
+)
+
+################################################################################
+NAME     = meson.project_name()
+SPECFILE = '@0@.spec'.format(NAME)
+
+substs = configuration_data()
```
Comment (verschwindibussen) [inline meson.build:189]: Please use the pkgconfig module from meson.

File: meson_options.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# -*- mode: meson -*-
+
+option('pkgconfiglibdir', type : 'string', value : '', description : 'directory for standard pkg-config files')
+
+option('systemd', type : 'combo', choices : ['auto', 'true', 'false'], description : 'libsystemd support')
```
Comment (verschwindibussen) [inline meson_options.txt:5]: Not completely sure about the combo type. Do we really want the auto?

File: pynvme/nvme.i [hunk 4]
```diff
@@ -377,7 +377,7 @@ struct nvme_ns {
 
 %extend nvme_host {
   nvme_host(struct nvme_root *r, const char *hostnqn = NULL,
-	    const char *hostid = NULL) {
+            const char *hostid = NULL) {
```
Comment (verschwindibussen) [inline pynvme/nvme.i:380]: Looks unrelated, please remove.

File: meson.build [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+substs.set('NAME',       NAME)
+substs.set('VERSION',    run_command(awk, '/Version:/ { print $2 }', SPECFILE).stdout().strip())
```
Comment (verschwindibussen) [inline meson.build:191]: Could we get rid of awk here. Reading this from the specfile looks a bit awkward.

**Review by verschwindibussen — COMMENTED** [review]
Generally looks good. A few small cleanup suggestions.

File: meson_options.txt (hunk 3 above)
Comment (martin-belanger) [reply meson_options.txt:5]: This is to automatically use these libraries if they are already installed. Otherwise, these features are automatically turned off.

Comment (verschwindibussen) [reply meson_options.txt:5]: We just discussed this here at ALPSS (we = Keith, Hannes and me). So we decided to make libuuid, json-c mandatory and get rig of the systemd dependency by implementing the uuid lookup directly in the library.

File: meson.build (hunk 5 above)
Comment (verschwindibussen) [reply meson.build:191]: Yes, we should avoid having two places defining the version. Hmm, not sure if we should remove the spec file.
